ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp

```diff
@@ -558,37 +558,51 @@ int get_color(int defcolor, bool advanced)
 
 string get_directory(string dname, string caption)
 {
-//NOTE: This uses the Windows Vista or later file chooser, which is different than the one used by GM8 and lower
-//because I could not find out which one it uses, since IFileDialog is used by both wxWidgets and QtFramework
-//and there doesn't appear to be a standard file picker for XP or lower in the Windows API except SHBrowseForFolder that is
-//used by Game Maker for get_directory_alt
-  IFileDialog* fileDialog;
-  CoCreateInstance(CLSID_FileOpenDialog, NULL, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&fileDialog));
+  //NOTE: This uses the Windows Vista or later file chooser, which is different than the one used by GM8 and lower
+  //because I could not find out which one it uses, since IFileDialog is used by both wxWidgets and QtFramework
+  //and there doesn't appear to be a standard file picker for XP or lower in the Windows API except SHBrowseForFolder that is
+  //used by Game Maker for get_directory_alt
+  IFileDialog *SelectDirectory;
+  CoCreateInstance(CLSID_FileOpenDialog, NULL, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&SelectDirectory));
 
   DWORD options;
-  fileDialog->GetOptions(&options);
-  options &= ~FOS_FILEMUSTEXIST;
-  options &= ~FOS_PATHMUSTEXIST;
```

Comment:
Are you sure GM doesn't allow the path or file to not exist? This seems like I would have tested this when in 32f553b4f2c07f31e359845b2b71f7c5a518696b I originally added these flags.

Reply:
Yes, I am positive. No, you must have not tested properly. Did you really think a normal GM user would benefit from a CLSID that non-paths return?

Reply:
No, I didn't, because I didn't know what a CLSID is because it genuinely sounds like Microsoft nonsense. I have never seen a file dialog that "selects COM objects" and while the idea doesn't sound far fetched it's totally insane. I blame them on poor API design more than myself not knowing what CLSID was. You have to understand the Microsoft document was complete garbage back in 2013 and didn't even provide code examples. The information you are finding now is easier for you because Microsoft has changed and is making a concerted effort to provide easily accessible Win32 documentation and such.

Reply:
The only reason I discovered the CLSID was from my testing, had nothing to do with reading the docs.

Reply:
Regardless, it was still the best code at the time because nobody had made any attempt to add the dialog until I did. Let me back up for a second though and clarify that I am not actually asking about the CLSID. Here's the source of confusion and maybe why I removed these options originally.

I ran this little test in GM8.1 just now.
```gml
show_message(get_directory("bleh"));
show_message(get_directory_alt("ok", "bleh"));
show_message(get_open_filename("", "bleh"));
```

The third file selecting dialog I think may have been implemented in ENIGMA. In ENIGMA and GM8.1 it allows you to select files and paths that don't exist by typing the name in the file dialog and just hitting open. As you can see below, using GM8.1's open file dialog I am able to open a file that doesn't exist.

![GM8 Open File Dialog Nonexistant](https://user-images.githubusercontent.com/3212801/54866468-68a3e180-4d4a-11e9-8a7b-5bc6ae5ad1f4.png)

That does not hold true for GM8.1's get directory functions. Both of the dialogs seem outdated compared to the file open dialog and do not allow you to actually type in the name of the dialog. Correct me if I am wrong, but I don't believe ENIGMA is using these outdated dialogs for selecting directories except in one case, I believe we use the newer variant? If that's true, and the user is able to type the name of the directory in, then I think it should allow selecting directories which don't exist to be kept consistent with the open file dialog. However, that does not mean I think it should be able to select COM objects.

I don't have anything else to add to this without more information I mentioned above.

Reply:
The thing is, if you allow selecting non existent locations, certain locations that do not count as valid directories will return a CLSID, no way around it. In the case of the file dialogs, I find it a bit silly you can select non-existent files, that defeats the purpose of selecting an file to begin with; the file must exist in order to be opened and processed. You sure you aren't testing the save dialog instead?

Reply:
Like I said you can try it yourself in 8.1. I will say that's a good distinction you made though. If, hypothetically, it's not a bug in GM then YoYo's logic may have been that it was intended so you could select a file you want to create. But that doesn't make sense, since that's what the save dialog would be for.
```cpp
show_message(get_open_filename("", "bleh"));
```

Reply:
Oh, I just realized, this is not resolved. I want you to consider that there is no such thing as a `get_save_directory` @time-killer-games, so how would you let your user select a directory that they want you to create? For example, what if it was about selecting a directory where you would like your game saved, and the game creates the directory if it doesn't exist? If you don't like it being able to select non-existant directories, we can make it a parameter and give it a default argument where it doesn't let you. Then we give people options. Again, I won't force you to do it, but now I see a reason to.

Reply:
Actually, it is. You can right click -> new folder in the dialog just like you can with get_directory_alt and the file dialogs.

Reply:
Yeah I was aware of that. I guess it really doesn't matter then. There's no reason to put in the extra work to see if the CLSID came from the file system. I guess we're good on this.